let $ c_{i \ j}$ be  The cost of constructing (setup cost) the $i^{th}$ facility at the $j^{th} $ location and $x_{i \ j} $ be the indicator variable 

i.e $x_{i \ j}$ = 1 if $i^{th}$ facility located at the $j^{th} $ location $x_{i \ j}$ = 0 if $i^{th}$ facility is not  located at the $j^{th} $ location 

let we have n facility and n location to assign

so,our objective function is 

 min$ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} x_{i \ j}*c_{i \ j}$

 subjected to $\sum_{i=1}^{i=n} x_{i \ j} =1 \ \ ∀j$ 
 
 i.e at a location we have to assign only one facility 

 $\sum_{j=1}^{j=n} x_{i \ j} =1 \ \  \ \ ∀i $ 
 
 i.e a  facility can be assighned at single  location 

 here i=(1,2,3,4,....................,n)

j=(1,2,3,4,...................,n)

$x_{i \ j} \in$ {0,1}

In [ ]:
!pip install -q pyomo

In [ ]:
import numpy as np

In [ ]:
from pyomo.environ import *

In [ ]:
import random as rd

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc=SolverFactory('cbc')

In [ ]:
def pyomo(data):
  global model_lab6_ex2
  model_lab6_ex2=ConcreteModel()
  M=len(data[:,0])
  N=len(data[0,:])
  global row_indices
  row_indices=np.arange(M)
  global col_indices
  col_indices=np.arange(N)
  model_lab6_ex2.x=Var(row_indices,col_indices,domain=Binary)
  model_lab6_ex2.objective=Objective(expr=sum(model_lab6_ex2.x[i,j]*data[i,j] for i in row_indices for j in col_indices),sense= minimize)
  model_lab6_ex2.constraints=ConstraintList()
  for i in row_indices:
      model_lab6_ex2.constraints.add(expr=sum(model_lab6_ex2.x[i,j] for j in col_indices) ==1)
  for i in col_indices:
     model_lab6_ex2.constraints.add(expr=sum(model_lab6_ex2.x[j,i] for j in row_indices) ==1)  

In [ ]:
cost_matrix=np.loadtxt('lab6_ex2.txt',delimiter='   ')

In [ ]:
pyomo(cost_matrix)

In [ ]:
  result = opt_cbc.solve(model_lab6_ex2)
  print('Solver status:', result.solver.status)
  print('Solver termination condition:',result.solver.termination_condition)
  # display solution
  print('\nObjective = ', model_lab6_ex2.objective())

  print('\nDecision Variables')
  for i in row_indices:
      for j in col_indices:
          if model_lab6_ex2.x[i,j].value !=0:
               print('Factory',i+1,'is located to',j+1,'location')


Solver status: ok
Solver termination condition: optimal

Objective =  204.0

Decision Variables
Factory 1 is located to 11 location
Factory 2 is located to 9 location
Factory 3 is located to 8 location
Factory 4 is located to 6 location
Factory 5 is located to 3 location
Factory 6 is located to 12 location
Factory 7 is located to 7 location
Factory 8 is located to 4 location
Factory 9 is located to 10 location
Factory 10 is located to 1 location
Factory 11 is located to 2 location
Factory 12 is located to 5 location


$\huge{solution \ \ of  \ \ 2.8}$

cost =  204.0

Factory 1 is located to 11 location

Factory 2 is located to 9 location

Factory 3 is located to 8 location

Factory 4 is located to 6 location

Factory 5 is located to 3 location

Factory 6 is located to 12 location

Factory 7 is located to 7 location

Factory 8 is located to 4 location

Factory 9 is located to 10 location

Factory 10 is located to 1 location

Factory 11 is located to 2 location

Factory 12 is located to 5 location

In [ ]:
model_lab6_ex2.x.domain=Reals
model_lab6_ex2.x.setlb(0)
model_lab6_ex2.x.setub(1)

In [ ]:
  result = opt_cbc.solve(model_lab6_ex2)
  print('Solver status:', result.solver.status)
  print('Solver termination condition:',result.solver.termination_condition)
  # display solution
  print('\nObjective = ', model_lab6_ex2.objective())

  print('\nDecision Variables')
  for i in row_indices:
      for j in col_indices:
          if model_lab6_ex2.x[i,j].value !=0:
               print('Factory',i+1,'is located to',j+1,'location')


Solver status: ok
Solver termination condition: optimal

Objective =  204.0

Decision Variables
Factory 1 is located to 11 location
Factory 2 is located to 9 location
Factory 3 is located to 8 location
Factory 4 is located to 6 location
Factory 5 is located to 3 location
Factory 6 is located to 12 location
Factory 7 is located to 7 location
Factory 8 is located to 4 location
Factory 9 is located to 10 location
Factory 10 is located to 1 location
Factory 11 is located to 2 location
Factory 12 is located to 5 location


$\huge{solution \ \ of  \ \ 2.9}$

cost =  204.0

Factory 1 is located to 11 location

Factory 2 is located to 9 location

Factory 3 is located to 8 location

Factory 4 is located to 6 location

Factory 5 is located to 3 location

Factory 6 is located to 12 location

Factory 7 is located to 7 location

Factory 8 is located to 4 location

Factory 9 is located to 10 location

Factory 10 is located to 1 location

Factory 11 is located to 2 location

Factory 12 is located to 5 location

$\huge{solution \ \ of \ \ 2.10}$

The MILP is bounded by some polyhedron with the additional constraint that only the integral points are feasible. The LP MODEL  is the same polyhedron where all points are feasible. Therefore, the first condition is satisfied as the set of all points in a polyhedron must logically include all integral points in the same polyhedron

now,if we  solve LP  and all variables are integral. Either (MILP) has the integrality property (where the optimal solution of the corresponding LP will always be integral, such as with LPs with totally unimodular matrices), or we were lucky and the optimal solution of MILP touches the edge of the polyhedron that bounds the feasible region

we can be certain that (for a minimisation problem), the optimal value to the corresponding MILP must be at least the optimal solution to LP, since the set of feasible points of MILP is contained inside that of LP, and because the function values of MILP at any arbitrary point is at least as great as that of LP at the same point.

i.e if soltion of LP is integer lies in the feasible reason of MILP then both solution are equal i.e feasible solution for the MILP is a feasible solution to the LP, and the reversed is not always so, i.e. a feasible solution for the LP is not necessarily a feasible solution for the MILP



In [ ]:
def result():
  result = opt_cbc.solve(model_lab6_ex2)
  # display solution
  print('\nObjective = ', model_lab6_ex2.objective())

  print('\nDecision Variables')
  for i in row_indices:
      for j in col_indices:
        if model_lab6_ex2.x[i,j].value!=0:
               print('x',i,j,'=',model_lab6_ex2.x[i,j].value)



In [ ]:
def result1():
  result = opt_cbc.solve(model_lab6_ex2)
  for i in row_indices:
      for j in col_indices:
        if model_lab6_ex2.x[i,j].value==float:
               print('x',i,j,'=',model_lab6_ex2.x[i,j].value)

In [ ]:
'''for i in range(10000):
  c=rd.sample(range(12),1)
  d=rd.sample(range(12),1)
  e=rd.uniform(0,1)
  if e<=0.5:
    f=40*e
  else:
    f=20*e
  print(c,d,f)
  m=cost_matrix[c[0],d[0]]
  cost_matrix[c[0],d[0]]=float(f)
  pyomo(cost_matrix)
  model_lab6_ex2.x.domain=Reals
  model_lab6_ex2.x.setlb(0)
  model_lab6_ex2.x.setub(1)
  cost_matrix[c[0],d[0]]=m
  result1())'''

'for i in range(10000):\n  c=rd.sample(range(12),1)\n  d=rd.sample(range(12),1)\n  e=rd.uniform(0,1)\n  if e<=0.5:\n    f=40*e\n  else:\n    f=20*e\n  print(c,d,f)\n  m=cost_matrix[c[0],d[0]]\n  cost_matrix[c[0],d[0]]=float(f)\n  pyomo(cost_matrix)\n  model_lab6_ex2.x.domain=Reals\n  model_lab6_ex2.x.setlb(0)\n  model_lab6_ex2.x.setub(1)\n  cost_matrix[c[0],d[0]]=m\n  result1())'

In [ ]:
for i in range(10):
  c=rd.sample(range(12),1)
  d=rd.sample(range(12),1)
  e=rd.uniform(0,1)
  if e<=0.5:
    f=40*e
  else:
    f=20*e
  print(c,d,f)
  m=cost_matrix[c[0],d[0]]
  cost_matrix[c[0],d[0]]=float(f)
  pyomo(cost_matrix)
  model_lab6_ex2.x.domain=Reals
  model_lab6_ex2.x.setlb(0)
  model_lab6_ex2.x.setub(1)
  cost_matrix[c[0],d[0]]=m
  result()




[1] [8] 12.589844070961165

Objective =  200.58984407096116

Decision Variables
x 0 10 = 1.0
x 1 8 = 1.0
x 2 7 = 1.0
x 3 5 = 1.0
x 4 2 = 1.0
x 5 11 = 1.0
x 6 6 = 1.0
x 7 3 = 1.0
x 8 9 = 1.0
x 9 0 = 1.0
x 10 1 = 1.0
x 11 4 = 1.0
[9] [8] 15.34923632258456

Objective =  203.34923632258455

Decision Variables
x 0 11 = 1.0
x 1 10 = 1.0
x 2 7 = 1.0
x 3 0 = 1.0
x 4 2 = 1.0
x 5 5 = 1.0
x 6 6 = 1.0
x 7 3 = 1.0
x 8 9 = 1.0
x 9 8 = 1.0
x 10 1 = 1.0
x 11 4 = 1.0
[3] [7] 17.856528337139757

Objective =  204.0

Decision Variables
x 0 10 = 1.0
x 1 8 = 1.0
x 2 7 = 1.0
x 3 5 = 1.0
x 4 2 = 1.0
x 5 11 = 1.0
x 6 6 = 1.0
x 7 3 = 1.0
x 8 9 = 1.0
x 9 0 = 1.0
x 10 1 = 1.0
x 11 4 = 1.0
[6] [11] 17.42767008432199

Objective =  200.427670084322

Decision Variables
x 0 10 = 1.0
x 1 8 = 1.0
x 2 6 = 1.0
x 3 4 = 1.0
x 4 2 = 1.0
x 5 5 = 1.0
x 6 11 = 1.0
x 7 3 = 1.0
x 8 9 = 1.0
x 9 0 = 1.0
x 10 1 = 1.0
x 11 7 = 1.0
[5] [10] 5.052546273638532

Objective =  192.05254627363854

Decision Variables
x 0 11 = 1.0
x 1 8 = 1.0

$\huge{solution \ \ of \ \ 2.12}$

since 3 can't assign to location 4  
 
  9 can't assign to location 6
 
  4 can't assign to laction 10

  for this we can changed  the text file by increasing the corresponding facility and location cost i.e if we take corresponding facility and location cost sufficiently large then our pyomo model doesn't assign that location since  we have to minimize cost and cost of corresponding location is too high 


  since in optimize solution it doesn't assign  (3,4)(9,6)(4,10) because it cost is already sufficientle large (it doesn't effect the soltion) so we have no need to change in code  we can see by below code 

In [ ]:
cost_matrix[2,3]=10000000
cost_matrix[8,5]=10000000
cost_matrix[3,9]=10000000

In [ ]:
model_lab6_ex2.x.setub(np.inf)
model_lab6_ex2.x.setlb(-np.inf)
model_lab6_ex2.x.domain=Binary

In [ ]:
  result = opt_cbc.solve(model_lab6_ex2)
  print('Solver status:', result.solver.status)
  print('Solver termination condition:',result.solver.termination_condition)
  # display solution
  print('\nObjective = ', model_lab6_ex2.objective())

  print('\nDecision Variables')
  for i in row_indices:
      for j in col_indices:
          if model_lab6_ex2.x[i,j].value !=0:
               print('Factory',i+1,'is located to',j+1,'location')

  print('\nConstraints')
  model_lab6_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  204.0

Decision Variables
Factory 1 is located to 11 location
Factory 2 is located to 9 location
Factory 3 is located to 8 location
Factory 4 is located to 6 location
Factory 5 is located to 3 location
Factory 6 is located to 12 location
Factory 7 is located to 7 location
Factory 8 is located to 4 location
Factory 9 is located to 10 location
Factory 10 is located to 1 location
Factory 11 is located to 2 location
Factory 12 is located to 5 location

Constraints
constraints : Size=24
    Key : Lower : Body : Upper
      1 :   1.0 :  1.0 :   1.0
      2 :   1.0 :  1.0 :   1.0
      3 :   1.0 :  1.0 :   1.0
      4 :   1.0 :  1.0 :   1.0
      5 :   1.0 :  1.0 :   1.0
      6 :   1.0 :  1.0 :   1.0
      7 :   1.0 :  1.0 :   1.0
      8 :   1.0 :  1.0 :   1.0
      9 :   1.0 :  1.0 :   1.0
     10 :   1.0 :  1.0 :   1.0
     11 :   1.0 :  1.0 :   1.0
     12 :   1.0 :  1.0 :   1.0
     13 :   1.0 :  1.0 :   1.0
     14 :

$\huge{solution \ \  of \ \ 2.12}$

cost =204

Factory 1 is located to 11 location

Factory 2 is located to 9 location

Factory 3 is located to 8 location

Factory 4 is located to 6 location

Factory 5 is located to 3 location

Factory 6 is located to 12 location

Factory 7 is located to 7 location

Factory 8 is located to 4 location

Factory 9 is located to 10 location

Factory 10 is located to 1 location

Factory 11 is located to 2 location

Factory 12 is located to 5 location

$\huge{observation}$

since in optimize solution it doesn't assign  (3,4)(9,6)(4,10) because it cost is already sufficientle large (it doesn't effect the soltion) so we have no need to change in code 

our complete solution is same as above this situation doesn't effect it